src/windows/types.rs

```diff
@@ -0,0 +1,303 @@
+        // For information:
+        //  - msvc-demangler has no problem with symbols containing ".llvm."
+
+        let name = Name::new(ident);
+        match name.demangle(DemangleOptions {
+            format: DemangleFormat::Full,
+            with_arguments: true,
+        }) {
+            Some(demangled) => demangled,
+            None => {
+                let lang = name.detect_language();
+                if lang != Language::Unknown {
+                    warn!("Didn't manage to demangle {}", ident);
+                }
+                ident.to_string()
+            }
+        }
+    }
+
+    fn dump_procedure_parts(&self, typ: ProcedureType) -> Result<(String, String)> {
+        let ret_typ = if let Some(ret_typ) = typ.return_type {
+            let attrs = typ.attributes;
+            if attrs.is_constructor() || attrs.cxx_return_udt() {
+                "".to_string()
+            } else {
+                self.dump_index(ret_typ)?
+            }
+        } else {
+            "".to_string()
+        };
+        let args_typ = self.dump_index(typ.argument_list)?;
+
+        Ok((ret_typ, args_typ))
+    }
+
+    fn check_this_type(&self, this: TypeIndex, class: TypeIndex) -> Result<bool> {
+        let this = self.finder.find(this)?;
+        let this = this.parse()?;
+        Ok(if let TypeData::Pointer(this) = this {
+            this.underlying_type == class
+        } else {
+            false
+        })
+    }
+
+    fn dump_method_parts(&self, typ: MemberFunctionType) -> Result<(String, String)> {
+        let attrs = typ.attributes;
+        let ret_typ = if attrs.is_constructor() || attrs.cxx_return_udt() {
+            "".to_string()
+        } else {
+            self.dump_index(typ.return_type)?
+        };
+        let args_typ = self.dump_index(typ.argument_list)?;
+        // Note: "this" isn't dumped but there are some cases in rust code where
+        // a first argument shouldn't be "this" but in fact it is:
+        // https://hg.mozilla.org/releases/mozilla-release/annotate/7ece03f6971968eede29275477502309bbe399da/toolkit/components/bitsdownload/src/bits_interface/task/service_task.rs#l217
+        // So we dump "this" when the underlying type (modulo pointer) is different from the class type
+
+        let args_typ = if let Some(this_typ) = typ.this_pointer_type {
+            if !self.check_this_type(this_typ, typ.class_type)? {
+                let this_typ = self.dump_index(this_typ)?;
+                if args_typ.is_empty() {
+                    this_typ
+                } else {
+                    format!("{}, {}", this_typ, args_typ)
+                }
+            } else {
+                args_typ
+            }
+        } else {
+            args_typ
+        };
+
+        Ok((ret_typ, args_typ))
+    }
+
+    fn dump_attributes(attrs: Vec<PointerAttributes>) -> String {
+        let mut buf = String::new();
+        if let Some((first, attrs)) = attrs.split_first() {
```

Comment:
To avoid the duplication here you can `map()` the elements of `attrs` to generate the strings for each one then `join()` them using the space character as the separator.

Reply:
if attributes are just [ptr,ptr] then the dump is "**" but if it's [ptr, const ptr] then we'll have "* const *" so in the first case I should join on "" and in the second on " ". So I don't think it's possible.
Or maybe we could have "* *" for the first, wdyt ?

Reply:
Good point. Well, you could just use the second loop then which would leave a space at the beginning if starting with a "const" and then `str::trim()` the result to get rid of it. So you wouldn't need a dedicated block just to handle the first element.